lib/ansible/modules/cloud/vmware/vmware_guest_tools_upgrade.py

```diff
@@ -0,0 +1,184 @@
+from ansible.module_utils._text import to_native
+from ansible.module_utils.vmware import (PyVmomi, gather_vm_facts, vmware_argument_spec,
+                                         wait_for_task)
+
+
+class PyVmomiHelper(PyVmomi):
+    def __init__(self, module):
+        super(PyVmomiHelper, self).__init__(module)
+
+    def upgrade_tools(self, vm):
+        result = {'failed': False, 'changed': False, 'msg': ''}
+        # Exit if VMware tools is already up to date
+        if vm.guest.toolsStatus == "toolsOk":
+            result.update(
+                changed=False,
+                msg="VMware tools is already up to date",
+            )
+            return result
+
+        # Fail if VM is not powered on
+        elif vm.summary.runtime.powerState != "poweredOn":
+            result.update(
+                failed=True,
+                msg="VM must be powered on to upgrade tools",
+            )
+            return result
+
+        # Fail if VMware tools is either not running or not installed
+        elif vm.guest.toolsStatus in ["toolsNotRunning", "toolsNotInstalled"]:
+            result.update(
+                failed=True,
+                msg="VMware tools is either not running or not installed",
+            )
+            return result
+
+        # If vmware tools is out of date, check major OS family
+        # Upgrade tools on Linux and Windows guests
+        elif vm.guest.toolsStatus == "toolsOld":
+            try:
+                if vm.guest.guestFamily in ["linuxGuest", "windowsGuest"]:
+                    task = vm.UpgradeTools()
+                wait_for_task(task)
+                result.update(
+                    changed=True,
+                )
+                return result
+            except Exception as exc:
+                result.update(
+                    failed=True,
+                    msg='Error while upgrading VMware tools %s' % exc,
+                )
+                return result
+        else:
+            result.update(
+                failed=True,
+                msg="VMWare tools could not be upgraded",
+            )
+            return result
+
+
+def main():
+    argument_spec = vmware_argument_spec()
+    argument_spec.update(
+        name=dict(type='str'),
+        name_match=dict(type='str', choices=['first', 'last'], default='first'),
+        uuid=dict(type='str'),
+        folder=dict(type='str', default='/vm'),
+        datacenter=dict(type='str', required=True),
+    )
+    module = AnsibleModule(argument_spec=argument_spec,
+                           required_one_of=[['name', 'uuid']])
+
+    if module.params['folder']:
+        # FindByInventoryPath() does not require an absolute path
+        # so we should leave the input folder path unmodified
+        module.params['folder'] = module.params['folder'].rstrip('/')
+
+    pyv = PyVmomiHelper(module)
+    # Check if the VM exists before continuing
+    vm = pyv.get_vm()
+
+    # VM already exists
+    if vm:
```

Comment:
Add another else for if we are not able to find the virtual machine providing appropriate error message.